rbx_types/src/attributes/reader.rs

```diff
@@ -0,0 +1,180 @@
+// Thanks to Anaminus! https://github.com/RobloxAPI/rbxattr/blob/master/spec.md
+
+use super::*;
+use crate::{basic_types::*, brick_color::BrickColor, variant::Variant};
+use std::{
+    collections::HashMap,
+    io::{self, Read},
+};
+
+fn read_u8<R: Read>(mut reader: R) -> io::Result<u8> {
+    let mut bytes = [0u8; 1];
+    reader.read_exact(&mut bytes)?;
+    Ok(bytes[0])
+}
+
+fn read_i32<R: Read>(mut reader: R) -> io::Result<i32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(i32::from_le_bytes(bytes))
+}
+
+fn read_u32<R: Read>(mut reader: R) -> io::Result<u32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(u32::from_le_bytes(bytes))
+}
+
+fn read_f32<R: Read>(mut reader: R) -> io::Result<f32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(f32::from_le_bytes(bytes))
+}
+
+fn read_string<R: Read>(mut reader: R) -> io::Result<String> {
```

Comment:
Given that strings in Lua don't necessarily have to be UTF-8, and Attributes support binary data (which won't be valid UTF-8), it seems like it might be pertinent to return the Vec instead of turning it into a String.

Reply:
Turns out this leads into a weird issue--all the AttributeTypes are `AttributeType::T => Variant::T`, it has made a lot of the code surrounding it easier to work with. In this case, `AttributeType::String => Variant::String`. Would you recommend making this BinaryString or similar?

Reply:
That's an unfortunate problem. It isn't an ideal solution but I'm going to hesitantly say "yes" to making it BinaryString since supporting binary data should be a priority of ours. Worst-case we can always change it later.